### ISSUE TITLE
Precision problem with stringValue in certain cases

### ISSUE THREAD

**Issue by nico75005** [davedelong/DDMathParser#99]
Hi, first thanks for this great library but I have a problem which actually seems well known with NSNumber.

When I evaluate a simple string through evaluateString like "1.2*6", the result is as expected 7.2 but when I call stringValue on this NSNumber(double), I get "7.199999999999999" or 7.1999999999999993 if I use doubleValue.

How can I do?

Thanks for your help

**Comment by nico75005** [comment]
Well I just saw you added usesHighPrecisionEvaluation. It seems to fix the issue with stringValue but doesn't fix it with doubleValue

**Comment by davedelong** [comment]
http://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html

Short of writing my own infinite-precision math framework, I'm not sure there's a way to "solve" this.

**Comment by JanX2** [comment]
What about porting one of the infinite-precision math framework written in C++? I have been toying with the idea every since you started the Swift branch. Someone just needs to find the time. ;)

**Comment by joshdistler** [comment]
I would love to see an infinite-precision implementation here. In testing, I'm seeing: 45.25 - 45 = 0.230000000000004. Any easy fix?

**Comment by davedelong** [comment]
No, this is not an easy fix. I'm waiting to implement this until Swift has a standardized "BigNum" library.

I've thought about making `Expression` generic with respect to its underlying number type, but so far there's too much up and down the stack that depends on having a concrete number.

**Comment by joshdistler** [comment]
Thanks Dave. Definitely looking forward to this very much as it's the only weakness in your otherwise amazing library. <fingers crossed>

**Comment by hani-ibrahim** [comment]
Hi @davedelong thanks for this amazing library 🙏 
Is there an update on this?
Can this be fixed if the output of the library becomes `Decimal` or `NSDecimalNumber` instead of `Double`?

**Comment by davedelong** [comment]
@hani-ibrahim 

There's no real update on this. DDMathParser used to have a `Decimal`-based evaluation engine, but the problem with that is that `Decimal` only has extremely rudimentary arithmetic primitives (addition and multiplication, essentially). Implementing things like the trigonometric functions became extremely complicated and error-prone, and I did not (and still do not) have the technical expertise to implement that correctly.

It wouldn't be hard to make the parser itself support a generic/abstract output type from its _parsing_, but I would only be able to supply the evaluation support for `Double`-based numbers.

A few more details are in this conversation I had with @uliwitness a month or so ago:

Uli's thread start: https://chaos.social/@uliwitness/114859653271948437
My answer: https://mastodon.social/@davedelong/114863586972874081

**Comment by hani-ibrahim** [comment]
Thanks a lot for the context @davedelong 🙏 
Very interesting....
